File: assignments/L2.py
#!/usr/bin/env python3

import json
import sys

data = json.load(sys.stdin)
for f in data["functions"]:
    n2b = {}
    blocks = []
    blk = []
    idx = 0
    for i in f["instrs"]:
        idx += 1
        op = i.get("op", "no_op")
        if op in ("br", "jmp"):
            blk.append(i)
            blocks.append(blk)
            blk = []
        elif op == "no_op": # this is a label
            if(len(blk) > 0):
                blocks.append(blk)
            blk = []
            blk.append(i)
            n2b[i["label"]] = blk
        else:
            blk.append(i)
            if(len(blk) == 1):
                n2b["B{}".format(idx)] = blk
    if(len(blk) > 0):
        blocks.append(blk)

    succs = {}
    for i in range(0, len(n2b)):
        name, b = list(n2b.items())[i]
        succs[name] = []
        last = b[len(b) - 1]
        if last["op"] in ("jmp", "br"):
            for l in last["labels"]:
                succs[name].append(l)
        else:
            if i < len(n2b) - 1:
                succs[name].append(list(n2b.keys())[i+1])
    print("digraph {} {{".format(f["name"]))
    for name in n2b:
        print("    {}".format(name))
    for name in succs:
        for l in succs[name]:
            print("    {} -> {}".format(name, l))
    print("}")

File: assignments/L2.cpp
#include <string>
#include <vector>
#include <iostream>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

class block {
private:
  std::string name;
  std::vector<json> instructions;

public:
  block() { name = "unamed"; }
  void set_name(int idx) {
    if (instructions[0].contains("label")) {
      name = instructions[0]["label"];
    } else {
      name = "B";
      name += to_string(idx);
    }
  }
  string get_name() { return name; }
  void append_instruction(json inst) { instructions.push_back(inst); }
  json get_last_instruction() { return instructions[size()-1];}
  size_t size() { return instructions.size();}
  void dump() {
    cout << "block: " << name << endl;
    for (auto inst : instructions)
      cout << inst << endl;
  }
};

int main() {
  json j;
  std::cin >> j;
  //std::cout << j.dump(4);
  for (auto func : j["functions"]) {
    std::vector<block*> blocks;
    std::map<std::string, block*> name2block;
    block *b = new block;
    for (auto inst : func["instrs"]) {
      // std::cout << inst << std::endl;
      if (inst.contains("op")) {
        b->append_instruction(inst);
        if (inst["op"] == "jmp" || inst["op"] == "br") {
          blocks.push_back(b);
          b = new block;
        }
      } else { // label
        if (b->size())
          blocks.push_back(b);
        b = new block;
        b->append_instruction(inst);
      }
    }
    if (b->size())
      blocks.push_back(b);
    else
      free(b);
#if 0
    for (auto b : blocks) {
      b->dump();
    }
#endif
    int i = 0;
    for (auto blk : blocks) {
      blk->set_name(i);
      name2block[blk->get_name()] = blk;
      i++;
    }
#if 0
    for (auto &[name, blk] : name2block) {
      std::cout << name << std::endl;
      blk->dump();
    }
#endif
#if 1
    // CFG
    std::map<block *, vector<block*>*> cfg;
    i = 0;
    for (auto &[name, blk] : name2block) {
      auto inst = blk->get_last_instruction();
      if (inst["op"] == "br" || inst["op"] == "jmp") {
        vector<block*> *blks = new vector<block*>();
        for (auto l : inst["labels"]) {
          blks->push_back(name2block[l]);
        }
        cfg[blk] = blks;
      } else {
        if (i < name2block.size() - 1) {
          vector<block*> *blks = new vector<block*>;
          blks->push_back(name2block[std::string("B") + std::to_string(i)]);
          cfg[blk] = blks;
        }
      }
      i++;
    }
    cout << "digraph {} {" << endl;
    for (auto [name, blk] : name2block) {
      cout << "    " << name << endl;
    }
    for (auto &[b, succs] : cfg) {
      for (auto s : *succs) {
        if(s)
          cout << "    " << b->get_name() << " -> " << s->get_name() << endl;
      }
    }
    cout << "}" << endl;
#endif
  }


  return 0;
}
